geotiepoints/viiinterpolator.py

```diff
@@ -0,0 +1,196 @@
+# THR_CARTESIAN: latitude limit above which interpolation is performed in cartesian coordinates
+THR_CARTESIAN = 60.
+# THR_USE_XY: z-coordinate limit above which latitude is computed from x and y coordinates
+THR_USE_XY = 0.8
+
+
+def tie_points_interpolation(data_on_tie_points, scan_alt_tie_points, tie_points_factor):
+    """Interpolate the data from the tie points to the pixel points.
+    The data are provided as a list of xarray DataArray objects, allowing to interpolate on several arrays
+    at the same time; however the individual arrays must have exactly the same dimensions.
+
+    Args:
+        data_on_tie_points: list of xarray DataArray objects containing the values defined on the tie points.
+        scan_alt_tie_points: number of tie points along the satellite track for each scan
+        tie_points_factor: sub-sampling factor of tie points wrt pixel points
+
+    Returns:
+        list of xarray DataArray objects containing the interpolated values on the pixel points.
+
+    """
+    # Extract the dimensions of the tie points array across and along track
+    n_tie_act, n_tie_alt = data_on_tie_points[0].shape
+    dim_act, dim_alt = data_on_tie_points[0].dims
+
+    # Check that the number of tie points along track is multiple of the number of tie points per scan
+    if n_tie_alt % scan_alt_tie_points != 0:
+        raise ValueError("The number of tie points in the along-route dimension must be a multiple of %d",
+                         scan_alt_tie_points)
+
+    # Compute the number of scans
+    n_scans = n_tie_alt // scan_alt_tie_points
+
+    # Compute the dimensions of the pixel points array across and along track
+    n_pixel_act = (n_tie_act - 1) * tie_points_factor
+    n_pixel_alt = (n_tie_alt - 1) * tie_points_factor
+
+    # Create the grids used for interpolation across the track
+    tie_grid_act = da.arange(0, n_pixel_act + 1, tie_points_factor)
+    pixels_grid_act = da.arange(0, n_pixel_act)
+
+    # Create the grids used for the interpolation along the track (must not include the spurious points between scans)
+    tie_grid_alt = da.arange(0, n_pixel_alt + 1, tie_points_factor)
+    n_pixel_alt_per_scan = (scan_alt_tie_points - 1) * tie_points_factor
+    pixel_grid_alt = iter(())
+    for j_scan in range(n_scans):
+        start_index_scan = j_scan * scan_alt_tie_points * tie_points_factor
+        pixel_grid_alt = chain(pixel_grid_alt, range(start_index_scan, start_index_scan + n_pixel_alt_per_scan))
+    pixel_grid_alt = list(pixel_grid_alt)
+
+    # Loop on all arrays
+    data_on_pixel_points = []
+    for data in data_on_tie_points:
+
+        if data.shape != (n_tie_act, n_tie_alt) or data.dims != (dim_act, dim_alt):
+            raise ValueError("The dimensions of the arrays are not consistent")
+
+        # Interpolate using the xarray interp function twice: first across, then along the scan
+        # (much faster than interpolating directly in the two dimensions)
```

Comment:
Chosen approach was based on an xarray issue. The issue is described here: https://github.com/pydata/xarray/issues/2223. The solution with two 1d interpolation instead of a single 2d interpolation is recommended by the developers. I see now that the issue has been closed (3 days ago!), so in principle using the latest xarray should allow to use directly the 2d interpolation, however the fix might not be available to all users. Results are exactly the same in both cases.